### PR TITLE
Feature/3.2/http routing

### PR DESCRIPTION
**Pull request by aliothe**
Makes it possible to define http service mappings for which tpcall will issue a http call instead.
This is something that can be preferable in a cloud environment.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -51,16 +51,42 @@ The following environment variables are expected
...
-  "transactionStickyEnabled": true
+  "transactionStickyEnabled": true,
+  "topologyChangeDelayMillis": 5000,
+  "routeFileName": "foo.json"
 }
 ```
 
+### HTTP Route Configuration
+
+If you have services that you want to be called non transactional via http then you can define them in a ```route.json``` file such as:
```
Comment (cklellie) [inline README.md]: transactionally

File: casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import static jakarta.ws.rs.client.Entity.entity;
+
+public class HttpClient implements Closeable
+{
+    private static final Logger LOG = Logger.getLogger(HttpClient.class.getName());
+    private final Client client;
+    public HttpClient()
+    {
+        // CDI public constructor
+        this.client = ClientBuilder.newClient();
```
Comment (cklellie) [inline casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java:33]: What happens when the client is closed and this bean is reused?
Should you use constructor based injection to ensure this is recreated/destroyed/closed using a producer instead with `@Produces` and `@Dispose`?

File: casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        Objects.requireNonNull(buffer, "buffer can not be null");
+        MediaType mediaType = CasualBufferTypeConverter.convert(CasualBufferType.unmarshall(buffer.getType()));
+        LOG.finest(() -> "issuing http request to " + uri);
+        Response response = client.target(uri).request(mediaType).post(entity(buffer.getBytes().get(0), mediaType));
+        ErrorState errorState = ResponseStatusConverter.convert(response.getStatusInfo().toEnum());
+        if (errorState != ErrorState.OK)
+        {
+            return errorResponse(errorState, response);
+        }
+        Optional<CasualBuffer> responseBuffer = ResponseConverter.convert(response);
```
Comment (cklellie) [inline casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java]: Why use `Optional` here? It is not an optional response, there is always a response buffer but sometimes it is empty.

If you have `ResponseConverter` return a `CasualBuffer` instead and when it was returning `Optional.empty()` before it will now return `ServiceBuffer.empty()`.
You can then remove the duplication of the `new ServiceReturn` in the following lines.
You have the exact same thing on `errorResponse` too.

File: casual/casual-caller-http-client/src/test/groovy/se/laz/casual/http/HttpClientTest.groovy [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+   @Shared
+   HttpClient client = HttpClient.of()
+   @Shared
+   def content = '{"msg":"bazinga!"}'
+   @Shared
+   def key = 'FLD_STRING1'
+   @Shared
+   def root = 'test'
+
+   JerseyTest resource = new JerseyTest(){
```
Comment (cklellie) [inline casual/casual-caller-http-client/src/test/groovy/se/laz/casual/http/HttpClientTest.groovy:38]: This seems like a lot of work to create a whole server implementation just to test the client.
Why not utilise RestClientDriver instead?
https://github.com/rest-driver/rest-driver

Just declare the rule:
```groovy
@Rule
public ClientDriverRule driver = new ClientDriverRule()
```
Then add your expectation in your `given`:
```groovy
        driver.addExpectation(
                onRequestTo( "/casual/servicename1" )
                        .withAnyParams()
                        .withBody( expectedJsonPayload, MediaType.APPLICATION_JSON )
                        .withMethod( ClientDriverRequest.Method.POST ),
                giveResponse( fakeJsonResponse, MediaType.APPLICATION_JSON )
        )
```
This uses assertions that the client actually invokes the expected http request with the expected data.
It then allows you to return the response data you wish or error codes, which can easliy be use with a `where` table:

```groovy
driver.addExpectation(
                onRequestTo( "/casual/servicename1"T )
                        .withAnyParams()
                        .withBody( toHalJson( workspace ), MediaType.APPLICATION_JSON )
                        .withMethod( ClientDriverRequest.Method.POST ),
                giveEmptyResponse().withStatus( status )
        )
```
It can also flag up if other http requests are made that are not expected.

File: casual/casual-caller-http-client/src/test/groovy/se/laz/casual/http/HttpClientTest.groovy [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+      given:
+      URI uri = new URI(resource.getBaseUri().toString() + "${root}/${TestPath.OK.path}")
+      ServiceReturn<CasualBuffer> response = client.request(uri, buffer)
+      expect:
+      response.getServiceReturnState() == returnState
+      response.getErrorState() == errorState
+      response.getReplyBuffer().getBytes() == buffer.getBytes()
+      where:
+      buffer                                                      || returnState                   || errorState
+      JsonBuffer.of(content)                                      || ServiceReturnState.TPSUCCESS  || ErrorState.OK
```
Comment (cklellie) [inline casual/casual-caller-http-client/src/test/groovy/se/laz/casual/http/HttpClientTest.groovy]: no need for `returnState` or `errorState` to be in the where table as they are fixed.
Same for all other tests below.

File: casual/casual-caller/src/main/java/se/laz/casual/connection/caller/CasualCallerImpl.java [hunk 6]
```diff
@@ -38,29 +42,36 @@ public class CasualCallerImpl implements CasualCaller
...
         this.transactionLess = transactionLess;
         this.failedDomainDiscoveryHandler = failedDomainDiscoveryHandler;
+        this.httpClient = httpClient;
         List<ConnectionFactoryEntry> possibleEntries = connectionFactoryProvider.get();
-        if(possibleEntries.isEmpty())
+        if (possibleEntries.isEmpty())
         {
             throw new CasualCallerException("No connection factories available, casual caller is not usable");
         }
+        this.serviceRoutes = ServiceRoutes.of(ConfigurationService.getInstance().getConfiguration());
```
Comment (cklellie) [inline casual/casual-caller/src/main/java/se/laz/casual/connection/caller/CasualCallerImpl.java]: Do we have the same fun here for configuration to fix later?

File: casual/casual-caller/src/main/java/se/laz/casual/connection/caller/CasualCallerImpl.java [hunk 7]
```diff
@@ -38,29 +42,36 @@ public class CasualCallerImpl implements CasualCaller
...
     }
 
     @Override
     public ServiceReturn<CasualBuffer> tpcall(String serviceName, CasualBuffer data, Flag<AtmiFlags> flags)
     {
         failedDomainDiscoveryHandler.issueDomainDiscoveryAndRepopulateCache();
-        return flags.isSet(AtmiFlags.TPNOTRAN) ? transactionLess.tpcall(() -> tpCaller.tpcall(serviceName, data, flags, lookup)) : tpCaller.tpcall(serviceName, data, flags, lookup);
+        return serviceRoutes.getRoute(serviceName).map(uri -> httpClient.request(uri, data))
```
Comment (cklellie) [inline casual/casual-caller/src/main/java/se/laz/casual/connection/caller/CasualCallerImpl.java:72]: This is getting difficult to read. Too much in a single chained method call.

File: casual/casual-caller/src/main/java/se/laz/casual/connection/caller/services/ServiceRoutes.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    private ServiceRoutes(Set<Route> routes)
+    {
+        this.routes = routes;
+    }
+    public static ServiceRoutes of(Configuration config)
+    {
+        ServiceRoutes serviceRoutes =  config.getRouteFileName()
+                                             .map(ServiceRouteReader::load)
+                                             .orElse(EMPTY);
+        // note: in case gson was used, it is lenient towards trailing comma leading to potential null values
```
Comment (cklellie) [inline casual/casual-caller/src/main/java/se/laz/casual/connection/caller/services/ServiceRoutes.java]: This should be a comment on the private method not here.

File: casual/casual-caller/src/test/groovy/se/laz/casual/connection/caller/CasualCallerImplTest.groovy [hunk 9]
```diff
@@ -24,6 +25,7 @@ class CasualCallerImplTest extends Specification
     CasualConnectionFactory fallBackConnectionFactory
     ConnectionFactoryEntry fallBackEntry
     TransactionLess transactionLess
+    HttpClient httpClient
```
Comment (cklellie) [inline casual/casual-caller/src/test/groovy/se/laz/casual/connection/caller/CasualCallerImplTest.groovy:31]: Where are the additional tests to check if the httpClient has been used or not? Additionally that tpCaller or transactionLess have not been called?
HttpClient should just be mocked here.
`0* httpClient._` etc etc

File: casual/casual-caller/src/main/java/se/laz/casual/connection/caller/config/Configuration.java [hunk 10]
```diff
@@ -141,34 +165,41 @@ public static class Builder
...
         private Long topologyChangeDelayMillis;
+        private String routeFilename;
 
         public Configuration build()
         {
             return new Configuration(this);
         }
 
-        public Builder jndiSearchRoot(String jndiSearchRoot)
+        public Builder withJndiSearchRoot(String jndiSearchRoot)
```
Comment (cklellie) [inline casual/casual-caller/src/main/java/se/laz/casual/connection/caller/config/Configuration.java:175]: why the change to `wither`?

File: README.md [hunk 11]
```diff
@@ -51,16 +51,42 @@ The following environment variables are expected
...
+{
+  "routes":
+  [
+    { "name": "test-service", "uri": "http://test-host.test:9990/test-service"},
+    { "name": "another-test-service", "uri": "http://test-host.test:9990/another-test-service"}
+  ]
+}
+```
+
+Using the example above, when a tpcall is issued to service `test-service`, instead of doing the tpcall a http request is dispatched to `http://test-host.test:9990/test-service`.
```
Comment (cklellie) [inline README.md:81]: I assume this relies on the http app you made, but you have not referenced that here?
Additionally, i would argue that a client for that should have been made within that repository and published, which can be utilised within this repository to reduce the work required here allowing you to just mock that imported client in this code.
I also feel we are making too many repositories with very tightly intertwinded dependencies to each other.

File: casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java (hunk 2 above)
Comment (aliothe) [reply casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java:33]: Nice catch, if the appserver keeps the injected ( into CasualCallerImpl) HttpClient arounds and does not create new instances - then that would be bad.
Is there any guarantee with CDI that it does it one way or the other?

File: casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java (hunk 3 above)
Comment (aliothe) [reply casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java]: Sure

File: casual/casual-caller-http-client/src/test/groovy/se/laz/casual/http/HttpClientTest.groovy (hunk 4 above)
Comment (aliothe) [reply casual/casual-caller-http-client/src/test/groovy/se/laz/casual/http/HttpClientTest.groovy:38]: Curious, why withAnyParams, they do not matter?
Usually you are extremely keen on matching parameters but not here?

Sure I can rewrite it usint ClientDriverRule, however - that would be a lot more work than letting it be as it is.

File: casual/casual-caller-http-client/src/test/groovy/se/laz/casual/http/HttpClientTest.groovy (hunk 5 above)
Comment (aliothe) [reply casual/casual-caller-http-client/src/test/groovy/se/laz/casual/http/HttpClientTest.groovy]: True

File: casual/casual-caller/src/main/java/se/laz/casual/connection/caller/config/Configuration.java (hunk 10 above)
Comment (aliothe) [reply casual/casual-caller/src/main/java/se/laz/casual/connection/caller/config/Configuration.java:175]: 2 reasons, 1 that it is what we are using in most if not all other builders.
Secondly, groovy got confused with function/attribute name being exactly the same.

File: README.md (hunk 11 above)
Comment (aliothe) [reply README.md:81]: No it does not, it has nothing to do with this at all.

Comment (aliothe) [reply README.md:81]: This code makes so that you can decide that tpcalls to wherever, casual or java appserver, should be made using http.

The http application *can* be used to expose such casual services from a java appserver if you so wish.

File: casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java (hunk 2 above)
Comment (aliothe) [reply casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java:33]:
```
If the bean is in @Dependent scope, then the client holds a direct reference to its instance. 

A new instance of the bean is bound to the life cycle of the newly created object. 

A bean in @Dependent scope is never shared between multiple injection points. 
```

File: casual/casual-caller-http-client/src/test/groovy/se/laz/casual/http/HttpClientTest.groovy (hunk 4 above)
Comment (cklellie) [reply casual/casual-caller-http-client/src/test/groovy/se/laz/casual/http/HttpClientTest.groovy:38]: Yes in mocks I don't like wildcards. These examples were just a quick copy paste from elsewhere, that I don't care what params are sent on a http level as they are not used at all. There are also methods for `withHeader`, `withoutHeader`. Use what is needed.

The reason I recommend this approach over the other is that it confirms the entire http interaction including the payload sent on the wire, which is not checked currently. 
If you don't call it with the right payload it will raise an exception during the test.

It shouldn't need to be that different total to the code you have now but it will improve readability.

File: casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java (hunk 2 above)
Comment (cklellie) [reply casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java:33]: It should be sufficient to just inject and use a producer with produces and dispose for the Client object. No need to do anything more.
You might want to even play around with "requestscoped" on the producer if there are any thread issues with the http client, but as you just pull a target out and read the response in a one liner i don't think you will have an issue in reality here.

File: README.md (hunk 11 above)
Comment (cklellie) [reply README.md:81]: I am lost. 
But, if you want to call via http a java app server service, it would need the new http application then, right?
Or is this only intended for calls from java to native?
I thought this was to provide the final symmetry for casual-java to http based tpcall requests? So you can call inbound or outbound through java.

File: casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java (hunk 2 above)
Comment (aliothe) [reply casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java:33]: The lifespan will be exactly the same according to the CDI documentation.
Sure, I can change the code but it will work *exactly* the same as it does now.

Comment (aliothe) [reply casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java:33]: The Jakarta http client is thread safe according to specification.

Comment (aliothe) [reply casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java:33]: Re producers, this is what they should be used for ( from the CDI spec):
```
A producer method acts as a source of objects to be injected, where:
• the objects to be injected are not required to be instances of beans, or
• the concrete type of the objects to be injected may vary at runtime, or
• the objects require some custom initialization that is not performed by the bean constructor
```

Comment (aliothe) [reply casual/casual-caller-http-client/src/main/java/se/laz/casual/http/HttpClient.java:33]:
```
6.4 Dependent pseudo-scope [dependent_context]
The @Dependent scope type is a pseudo-scope. Beans declared with scope type @Dependent behave differently to beans
with other built-in scope types.
When a bean is declared to have @Dependent scope:
• No injected instance of the bean is ever shared between multiple injection points.
• Any instance of the bean injected into an object that is being created by the container is bound to the lifecycle of the newly
created object.
• When a Unified EL expression in a JSF or JSP page that refers to the bean by its bean name is evaluated, at most one instance
of the bean is instantiated. This instance exists to service just a single evaluation of the EL expression. It is reused if the bean
name appears multiple times in the EL expression, but is never reused when the EL expression is evaluated again, or when
another EL expression is evaluated.
• Any instance of the bean that receives a producer method, producer field, disposer method or observer method invocation exists
to service that invocation only.
• Any instance of the bean injected into method parameters of a disposer method or observer method exists to service the method
invocation only (except for observer methods of container lifecycle events).
Every invocation of the get() operation of the Context object for the @Dependent scope with a CreationalContext
returns a new instance of the given bean.
Every invocation of the get() operation of the Context object for the @Dependent scope with no CreationalContext
returns a null value.
The @Dependent scope is always active.
```

**Review by cklellie — APPROVED** [review]
Add additional sentences to the documentation about the dependency upon usage of casual with nginx and casual java http and you're good to go.
